packages/core/src/types.ts

```diff
@@ -75,6 +81,365 @@ export interface PositionWithOffset extends Position {
  */
 export type SourceRange = [number, number];
 
+//------------------------------------------------------------------------------
+// Rules
+//------------------------------------------------------------------------------
+
+/**
+ * What the rule is responsible for finding:
+ * - `problem` means the rule has noticed a potential error.
+ * - `suggestion` means the rule suggests an alternate or better approach.
+ * - `layout` means the rule is looking at spacing, indentation, etc.
+ */
+export type RuleType = "problem" | "suggestion" | "layout";
+
+/**
+ * The type of fix the rule can provide:
+ * - `code` means the rule can fix syntax.
+ * - `whitespace` means the rule can fix spacing and indentation.
+ */
+export type RuleFixType = "code" | "whitespace";
+
+/**
+ * An object containing visitor information for a rule. Each method is either the
+ * name of a node type or a selector, or is a method that will be called at specific
+ * times during the traversal.
+ */
+export interface RuleVisitor {
+	/**
+	 * Called for each node in the AST or at specific times during the traversal.
+	 */
+	[key: string]:
+		| ((node: unknown, parent?: unknown) => void)
+		| ((...unknown: unknown[]) => void);
+}
+
+/**
+ * Rule meta information used for documentation.
+ */
+export interface RulesMetaDocs {
+	/**
+	 * A short description of the rule.
+	 */
+	description?: string | undefined;
+
+	/**
+	 * The URL to the documentation for the rule.
+	 */
+	url?: string | undefined;
+
+	/**
+	 * The category the rule falls under.
+	 * @deprecated No longer used.
+	 */
+	category?: string | undefined;
+
+	/**
+	 * Indicates if the rule is generally recommended for all users.
+	 */
+	recommended?: boolean | undefined;
+}
+
+/**
+ * Meta information about a rule.
+ */
+export interface RulesMeta {
```

Comment:
Similar to the note on `RuleTextEditor` below, `messages` and  could be made generic. We do this in typescript-eslint's types so that rule tests have a type error reported if an incorrect message ID is used.

For reference, https://github.com/typescript-eslint/typescript-eslint/blob/2ad3404690cd601d1bbb80daa7f861ab5bb127e4/packages/utils/src/eslint-utils/RuleCreator.ts#L32-L38:

```ts
export interface RuleWithMeta<
  Options extends readonly unknown[],
  MessageIds extends string,
  Docs = unknown,
> extends RuleCreateAndOptions<Options, MessageIds> {
  meta: RuleMetaData<MessageIds, Docs>;
}
```

That `Docs` generic is something to consider here too. Plugins can define their own docs, and we found often want to have their own properties on top of ESLint's. The `RuleMetadata` type for `docs` is an `&` intersection of the provided `Docs` and the ESLint standard `RuleMetaDataDocs`. https://typescript-eslint.io/blog/announcing-typescript-eslint-v8#custom-rule-metadocs-types

Reply:
I'll need to look at this a bit closer. The code snippet you pasted above makes my head hurt -- I have a hard time reasoning about what it actually means.

Reply:
Yeah we're not huge fans of it ourselves 😄. A general summary if it helps...

What's happening is that each `RuleWithMeta` has _three_ "generic" parts that can be substituted in:

* What `Options` the rule has specified, which _must be_**\*** a `readonly` array 
* The `MessageIds` it can report, which _must be_**\*** a `string`
* The shape of `Doc`umentation, which doesn't need to be explicitly specified - it defaults to `unknown`

**\***_"must be"_ here is "assignable to": as in, whatever type we provide must satisfy that type:
* `readonly unknown[]` can be satisfied by, say, `[]` or `[{ allowInterfaces: boolean; allowObjects: boolean; }]`
* `MessageIds` can be satisfied by, say, `"noEmptyInterface" | `"noEmptyObject"``

It then _`extends`_ the interface `RuleCreateAndOptions`, which defines the standard JS `create()` method and typescript-eslint's `defaultOptions`. Plus then it adds a `meta` object of type `RuleMetaData`.

Putting it all together, this is a rough subset of [`@typescript-eslint/no-empty-object-type`](https://typescript-eslint.io/rules/no-empty-object-type):

```ts
type NoEmptyObjectType = RuleWithMeta<
  // Options
  [
    {
      allowInterfaces: boolean;
      allowObjects: boolean;
    }
  ],

  // MessageIds
  "noEmptyInterface" | "noEmptyObject",

  // Docs
  {
    recommended: 'recommended';
    requiresTypeChecking: true;
  }
>
```